sklearn/linear_model/ridge.py

```diff
@@ -1087,55 +1386,43 @@ def fit(self, X, y, sample_weight=None):
             X, y, self.fit_intercept, self.normalize, self.copy_X,
             sample_weight=sample_weight)
 
-        gcv_mode = self.gcv_mode
-        with_sw = len(np.shape(sample_weight))
-
-        if gcv_mode is None or gcv_mode == 'auto':
-            if sparse.issparse(X) or n_features > n_samples or with_sw:
-                gcv_mode = 'eigen'
-            else:
-                gcv_mode = 'svd'
-        elif gcv_mode == "svd" and with_sw:
-            # FIXME non-uniform sample weights not yet supported
-            warnings.warn("non-uniform sample weights unsupported for svd, "
-                          "forcing usage of eigen")
-            gcv_mode = 'eigen'
+        gcv_mode = _check_gcv_mode(X, self.gcv_mode)
 
         if gcv_mode == 'eigen':
-            _pre_compute = self._pre_compute
-            _errors = self._errors
-            _values = self._values
+            _decompose = self._decompose_gram
+            _errors = self._errors_gram
+            _values = self._values_gram
         elif gcv_mode == 'svd':
             # assert n_samples >= n_features
-            _pre_compute = self._pre_compute_svd
-            _errors = self._errors_svd
-            _values = self._values_svd
-        else:
-            raise ValueError('bad gcv_mode "%s"' % gcv_mode)
+            if sparse.issparse(X):
+                _decompose = self._decompose_covariance_sparse
+                _errors = self._errors_covariance_sparse
+                _values = self._values_covariance_sparse
+            else:
+                _decompose = self._decompose_covariance_dense
+                _errors = self._errors_covariance_dense
+                _values = self._values_covariance_dense
 
         if sample_weight is not None:
             X, y = _rescale_data(X, y, sample_weight)
-
-        centered_kernel = not sparse.issparse(X) and self.fit_intercept
-
-        v, Q, QT_y = _pre_compute(X, y, centered_kernel)
+            sqrt_sw = np.sqrt(sample_weight)
+        else:
+            sqrt_sw = np.ones(X.shape[0], dtype=X.dtype)
         n_y = 1 if len(y.shape) == 1 else y.shape[1]
         cv_values = np.zeros((n_samples * n_y, len(self.alphas)))
         C = []
 
         scorer = check_scoring(self, scoring=self.scoring, allow_none=True)
```

Comment:
Nit: Moving the `check_scoring` before `n_y` to fail error, and have `C` and `cv_values` defined closer to where it will be used.